CRM/Upgrade/Incremental/php/TimezoneRevertTrait.php

```diff
@@ -0,0 +1,118 @@
+    }
+  }
+
+  /**
+   * dev/core#2122 - keep a copy of converted dates
+   * In theory we could skip this step if logging is enabled, but (a) people
+   * might turn off logging before running upgrades, and (b) there may not be a
+   * complete record anyway. People can drop the new column if they don't need
+   * it.
+   * @param \CRM_Queue_TaskContext $ctx
+   * @return bool
+   */
+  public static function fillBackupEventDates(CRM_Queue_TaskContext $ctx): bool {
+    // We only run if the field is timestamp, so don't need to check about that.
+    CRM_Core_DAO::executeQuery('UPDATE civicrm_event SET start_date_ts_bak = start_date, end_date_ts_bak = end_date, registration_start_date_ts_bak = registration_start_date, registration_end_date_ts_bak = registration_end_date');
+    CRM_Core_DAO::executeQuery("ALTER TABLE civicrm_event CHANGE COLUMN event_tz event_tz_bak text NULL DEFAULT NULL COMMENT 'For troubleshooting upgrades post 5.47. Can drop this column if no issues.'");
+    return TRUE;
+  }
+
+  /**
+   * dev/core#2122 - undo timestamp conversion from 5.47
+   * @param \CRM_Queue_TaskContext $ctx
+   * @return bool
+   */
+  public static function revertEventDates(CRM_Queue_TaskContext $ctx): bool {
+    // We only run if the field is timestamp, so don't need to check about that.
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_event` MODIFY COLUMN `start_date` datetime DEFAULT NULL COMMENT 'Date and time that event starts.'");
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_event` MODIFY COLUMN `end_date` datetime DEFAULT NULL COMMENT 'Date and time that event ends. May be NULL if no defined end date/time'");
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_event` MODIFY COLUMN `registration_start_date` datetime DEFAULT NULL COMMENT 'Date and time that online registration starts.'");
+    CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_event` MODIFY COLUMN `registration_end_date` datetime DEFAULT NULL COMMENT 'Date and time that online registration ends.'");
+    return TRUE;
+  }
+
+  /**
+   * Check if civicrm_event start_date is a timestamp.
+   * @return bool
+   */
+  private static function areEventsUsingTimestamp(): bool {
+    $dao = CRM_Core_DAO::executeQuery("SHOW COLUMNS FROM civicrm_event LIKE 'start_date'");
+    if ($dao->fetch()) {
+      return (strtolower($dao->Type) === 'timestamp');
+    }
+    return FALSE;
+  }
+
+  /**
+   * Are there any events in the system?
+   * @return bool
+   */
+  private static function areThereAnyCiviEvents(): bool {
+    return (bool) CRM_Core_DAO::singleValueQuery('SELECT COUNT(id) FROM civicrm_event');
+  }
+
+  /**
+   * Return the event_tz that is used most often. The idea being that if they
+   * didn't change too much after upgrading to 5.47 this will be the timezone
+   * of the account that was used to do the original upgrade to 5.47.
+   * @return string
+   */
+  private static function getMajorityTimezone(): string {
```

Comment:
I had removed this function when I shortened the message. Unless the point of this trait is that you're expecting this is going to be needed again elsewhere. Hopefully we're not going to regularly need timezone revert features in future.

Reply:
Yeah, apologies for the uncommented-push. 

I was `r-run`ning after your update yesterday and wanted to try tweaking the message again. A couple things coming out of that

(1) I found the messages were showing up twice because the code was duplicated in `FiveFortySeven` and `FiveFortyEight`. I figured this was an accident -- something easy to do when shuffling the patches back-and-forth between the two upgraders (`FiveFortySeven`/`FiveFortyEight`) and the two branches (`5.47`/`5.48`). The `TimezoneRevertTrait` can be an exact in both branches, so it should be easier to shuffle between them.

(2) I liked that the message was shorter, but it was funny - I would've trimmed the other way round. (eg I would've kept+trimmed different parts of your earlier text.) Giving some indication of the current CiviEvent seems useful. I think maybe it's better to use bullets or line-breaks (ie giving similar info - but in smaller chunks).

(3) Rather than get into literary analysis, I thought I'd play with the prose and send concrete suggestions. (But I went to bed before sending that up.) Will post that soon.